### ISSUE TITLE
Can boost::math::float_distance be sped up?

### ISSUE THREAD

**Issue by NAThompson** [boostorg/math#359]
In the AGM PR, I have found that ~90% of the runtime is spent computing float distances. However, at least for `float` and `double`, the following trivial modification drops the runtime to a negligible fraction of the total runtime:

```
    int32_t fast_float_distance(float x, float y) {
        static_assert(sizeof(float) == sizeof(int32_t), "float is incorrect size.");
        int32_t xi = *reinterpret_cast<int32_t*>(&x);
        int32_t yi = *reinterpret_cast<int32_t*>(&y);
        return yi - xi;
    }

    int64_t fast_float_distance(double x, double y) {
        static_assert(sizeof(double) == sizeof(int64_t), "double is incorrect size.");
        int64_t xi = *reinterpret_cast<int64_t*>(&x);
        int64_t yi = *reinterpret_cast<int64_t*>(&y);
        return yi - xi;
    }
```

It seems like `boost::math::float_distance` is considerably more general than this, but can we dive through a happy path to extract performance in the trivial cases?

**Comment by jzmaddock** [comment]
Thats interesting!  Does it pass the tests?

**Comment by NAThompson** [comment]
@jzmaddock : Yes; and although I can't find the reference now, I believe that this is a mandatory property of IEEE floating point numbers.

**Comment by jzmaddock** [comment]
OK, but as pointed out in the article, your trick fails when the two inputs differ in sign (this includes when one input is zero).  I also get a negative rather than positive result for say `fast_float_distance(-1, -0.5)`.  All of which can be fixed with some special case handling of course...

**Comment by NAThompson** [comment]
@jzmaddock : Yeah, the fact that `agm` requires positive numbers (over the reals) simplifies the logic considerably.

**Comment by cosurgi** [comment]
`long double` is quite irritating sometimes. However it's also useful ;)

**Comment by NAThompson** [comment]
Duplicate of #775
